tests/chainer_tests/link_hooks_tests/test_weight_standardization.py

```diff
@@ -0,0 +1,190 @@
+from chainer import testing
+from chainer.testing import attr
+
+
+class TestExceptions(unittest.TestCase):
+
+    def setUp(self):
+        self.x = chainer.Variable(numpy.ones((10, 5), dtype=numpy.float32))
+        self.layer = L.Linear(5, 20)
+
+    def test_wrong_weight_name(self):
+        wrong_Weight_name = 'w'
+        hook = WeightStandardization(weight_name=wrong_Weight_name)
+        with pytest.raises(ValueError):
+            self.layer.add_hook(hook)
+
+    def test_raises(self):
+        with pytest.raises(NotImplementedError):
+            with WeightStandardization():
+                self.layer(self.x)
+
+    def test_invalid_shaped_weight(self):
+        with pytest.raises(ValueError):
+            L.Linear(10, 0).add_hook(WeightStandardization())
+
+
+class BaseTest(object):
+
+    def test_add_ws_hook(self):
+        layer, hook = self.layer, self.hook
+        layer.add_hook(hook)
+        if self.lazy_init:
+            with chainer.using_config('train', False):
+                layer(self.x)
+        if not self.use_gamma:
+            assert not hasattr(layer, 'gamma')
+        else:  # Use gamma parameter
+            assert hasattr(layer, 'gamma')
+            assert layer.gamma.ndim == 0 and layer.gamma.size == 1
+
+    def _init_layer(self):
+        hook = WeightStandardization()
+        layer = self.layer
+        layer.add_hook(hook)
+        if self.lazy_init:
+            # Initialize weight and bias.
+            with chainer.using_config('train', False):
+                layer(self.x)
+        return layer, hook
+
+    def check_weight_is_parameter(self, gpu):
+        layer, hook = self._init_layer()
+        if gpu:
+            layer = layer.to_gpu()
+        source_weight = getattr(layer, hook.weight_name)
+        x = cuda.to_gpu(self.x) if gpu else self.x
+        layer(x)
+        assert getattr(layer, hook.weight_name) is source_weight
+
+    def test_weight_is_parameter_cpu(self):
+        if not self.lazy_init:
+            self.check_weight_is_parameter(False)
+
+    @attr.gpu
+    def test_weight_is_parameter_gpu(self):
+        if not self.lazy_init:
+            self.check_weight_is_parameter(True)
+
+    def check_deleted(self, gpu):
+        layer, hook = self.layer, self.hook
+        layer.add_hook(hook)
+        if gpu:
+            layer = layer.to_gpu()
+        x = cuda.to_gpu(self.x) if gpu else self.x
+
+        y1 = layer(x).array
+        with chainer.using_config('train', False):
+            y2 = layer(x).array
+        layer.delete_hook(hook.name)
+        y3 = layer(x).array
+        if gpu:
+            y1, y2, y3 = cuda.to_cpu(y1), cuda.to_cpu(y2), cuda.to_cpu(y3)
+        assert not numpy.array_equal(y1, y3)
+        assert not numpy.array_equal(y2, y3)
+
+    def test_deleted_cpu(self):
+        self.check_deleted(False)
+
+    @attr.gpu
+    def test_deleted_gpu(self):
+        self.check_deleted(True)
+
+
+class TestEmbedID(unittest.TestCase, BaseTest):
```

Comment:
I don't think this necessary because WeightStandardization is proposed for 2D convolutional layers followed by Batch/Group Normalization.

What do you think?